Fms2 io implementation

laurenchilutti opened the pull request:
This pull request implements fms2io into the GFDL_atmos_cubed_sphere code.  In a coming release of FMS, the modules ```mpp_io_mod``` and ```fms_io_mod``` are being deprecated and replaced by the new ```fms2_io_mod```.  

These changes have been tested with SHiELD and do not change answers.

lharris4 commented:
Hi, Lauren. Thanks. Previously we had seen significant slowdowns in SHiELD
when using FMS2io. Have these been fixed?

Lucas

On Thu, Mar 11, 2021 at 1:21 PM laurenchilutti ***@***.***>
wrote:

> This pull request implements fms2io into the GFDL_atmos_cubed_sphere code.
> In a coming release of FMS, the modules mpp_io_mod and fms_io_mod are
> being deprecated and replaced by the new fms2_io_mod.
>
> These changes have been tested with SHiELD and do not change answers.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74
> Commit Summary
>
>    - Removing use of mpp_io_mod and fms_io_mod from the dycore code.
>    Replacing the necessary functions with fms2_io_mod functions
>    - Adding a call to set_filename_appendix so that nest is added to
>    filename when needed and removing unneccessary code in fv_io_mod
>
> File Changes
>
>    - *M* GFDL_tools/fv_ada_nudge.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-5f2444cd4095f79e9ca9f0aae0a4ddead8d6cce49fcb9f82bd91783fd0410926>
>    (105)
>    - *M* GFDL_tools/fv_climate_nudge.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-25cf6f2e78899f71f2c0563cd1e5494f092e1fee52335a71bfddf9087d884834>
>    (18)
>    - *M* GFDL_tools/fv_cmip_diag.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-3f1b8ac017c0ebad7a0634f0b1d3936bc2010f34c7d9ed0259a5055af751e930>
>    (30)
>    - *M* GFDL_tools/fv_diag_column.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-ac38063693e5aa2005a9f36024a21e9307eeb8774c82ad78530932d86e7c4a36>
>    (8)
>    - *M* GFDL_tools/read_climate_nudge_data.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-0d6d296008e396837aeb5d653923be7c855bbb632cc75851f95ff6add22b8710>
>    (232)
>    - *M* driver/GFDL/atmosphere.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-7e8a1f6bb2b5bac34a551a12a1137bda85088edceecf3b8e264f57da04d86a5f>
>    (24)
>    - *M* driver/SHiELD/atmosphere.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-fef260349c9878099baa4ddb913f585051bcaec17b14f2804309c325d2265307>
>    (24)
>    - *M* model/fv_arrays.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-48d8298cb86a49d043dbf0dfdca8cebd9ce7f043f42d0c12c0e8358d235b4af4>
>    (33)
>    - *M* model/fv_control.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-24043aed6dae86f0c0b7f564e7d936ea9a3b3dad70622ff8eaa8457abd47f782>
>    (45)
>    - *M* model/fv_regional_bc.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-ae1eccb8eddb7b619f528b8499ca6710da67274417b924e86e99eb37311f40e1>
>    (151)
>    - *M* tools/coarse_grained_restart_files.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-b95d5ed0814da01c334e8469b787e87cfaa03f8898937b94e9fbbdcd48ff07ad>
>    (271)
>    - *M* tools/coarse_graining.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-dea163df53e70a90304e5ee27e842069404db5822faac4778c9b85ecf7c702ac>
>    (2)
>    - *M* tools/external_ic.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-478347b92b9eb0c6a53e0959daefb2178df066500f17c48c79204f4e6336b210>
>    (605)
>    - *M* tools/fv_diagnostics.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-219422707190eb595680ca8763441ce74e0886412aa0f60e820c07b1b8169020>
>    (9)
>    - *M* tools/fv_grid_tools.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-137dd6ce54e3f7ec3e5896243e2d2d72ea43e931a2840b58138808bb16303d0c>
>    (187)
>    - *M* tools/fv_io.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-9fed1b9af94f1ecc043b5d67c050c9a1b990705ce2f1dfa9364e8abbed6c54f0>
>    (844)
>    - *M* tools/fv_mp_mod.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-7aa04ec8e3330bc5192885d0e11026a28faae63629af698981baedff02dc89a0>
>    (6)
>    - *M* tools/fv_nggps_diag.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-b0e7732ed4bfa6bcace4abc8914b750533145e81a735543af3b7c5f679c9c40e>
>    (9)
>    - *M* tools/fv_nudge.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-fcc04cabd7f46ee5d4e4e25029090f57b407096c1fec950c60033d1cd191120d>
>    (20)
>    - *M* tools/fv_restart.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-b6fd68af0ee5eab265be1535c3d772adc5279c8c2ec86f16e02bdd404372265d>
>    (23)
>    - *M* tools/fv_surf_map.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-02af510a0d067e4b1dd16ea071e2fc67786f6fb8935406753ab17ad51ecfc44b>
>    (16)
>    - *M* tools/fv_treat_da_inc.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-dd6790510b0f63c9cb33d51c8a430dd6884a5da8de66b40198bea46b7fb802e6>
>    (5)
>    - *M* tools/test_cases.F90
>    <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74/files#diff-75be47ee6af9df0d9d4d49357613de323ab5cab0353b1941a3871e54f3fe8800>
>    (10)
>
> Patch Links:
>
>    - https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74.patch
>    - https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMUQRVAKCT5SW372UMM4DMLTDD3ZFANCNFSM4ZAY7JPQ>
> .
>

laurenchilutti commented:
Lucas - I can run some tests and compare the timing.

bensonr commented:
The performance degradation you are talking about was a fundamental issue
initially that has since been rectified.  In fact, anything you are running
today with a recent release of FMS (2020.03 and newer) has the diagnostics
being output via the fms2_io interfaces.  If you aren't seeing runtime
impacts at the end of a simulation when all diagnostic data is being pushed
to storage, it is unlikely to occur with the restarts.  Regardless, a few
timing tests will be run to prove it.

bensonr commented:
@laurenchilutti - This PR was opened prior to a template being in place.  There is a lot of good work here, but I want to make sure we've tested to the largest extent that we can.  Could you list the tests you've run with the fms2_io to give an idea of the code coverage?  In particular, it would be good to list the IC type (GFS, HRRR, ECMWF, SHiELD), model type (nested, regional, etc.) and whether restarts (intermediate writing and functional test) were a part of the testing as well.  Also, are there changes coming/needed within the [SHiELD_driver](https://github.com/NOAA-GFDL/SHiELD_driver) project?

bensonr reviewed (changes requested):
Can you remove any references to the use association of  write_version_number from fms_io_mod.  I think the only place it appears is in tools/fv_diagnostics.F90 where it is also use associated via fms_mod.

laurenchilutti commented:
@bensonr My apologies for not directly responding to the comment from 6 days ago.  I am putting a list of all tests run and will update the description of this PR to include that.  SHiELD_drivers will also require updates to fms2io that has not been finished yet.  For now, the branch: [user/lec/fms2io](https://github.com/laurenchilutti/SHiELD_driver/tree/user/lec%2Ffms2io) from my fork is needed to test fms2io with SHiELD.

I will go ahead and clean up the use of write_version_mod.

lharris4 commented:
Hi, Lauren. Thanks for working on this.

Lucas

On Wed, Apr 28, 2021 at 12:19 PM laurenchilutti ***@***.***>
wrote:

> @bensonr <https://github.com/bensonr> My apologies for not directly
> responding to the comment from 6 days ago. I am putting a list of all tests
> run and will update the description of this PR to include that.
> SHiELD_drivers will also require updates to fms2io that has not been
> finished yet. For now, the branch: user/lec/fms2io
> <https://github.com/laurenchilutti/SHiELD_driver/tree/user/lec%2Ffms2io>
> from my fork is needed to test fms2io with SHiELD.
>
> I will go ahead and clean up the use of write_version_mod.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/74#issuecomment-828589057>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMUQRVFI4GDZVDGM6KLQ2KLTLAYSHANCNFSM4ZAY7JPQ>
> .
>

laurenchilutti commented:
> Can you remove any references to the use association of write_version_number from fms_io_mod. I think the only place it appears is in tools/fv_diagnostics.F90 where it is also use associated via fms_mod.

@bensonr I do not see any use statements from fms_io_mod anywhere in this new code.  When you look at tools/fv_diagnostics.F90 it used to  use write_version_number from fms_io_mod, but I removed and replaced that with fms_mod.  I may be misunderstanding your request.

bensonr commented:
@laurenchilutti thanks for confirming.  I was looking at the [current code](https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/blob/master/tools/fv_diagnostics.F90#L30) and didn't realize you'd fixed it as part of this PR.

bensonr commented on tools/external_ic.F90:
```diff
@@ -396,22 +395,16 @@ subroutine get_nggps_ic (Atm, fv_domain)
-!--- set the 'nestXX' appendix for all files using fms_io
-    if (Atm%grid_number > 1) then
-       write(gn,'(A4, I2.2)') "nest", Atm%grid_number
+    fname = 'INPUT/'//trim(fn_gfs_ctl)
```
Is there a reason fn_gfs_ctl doesn't have INPUT/ prepended like the other fn_*_ic filenames?

laurenchilutti replied:
There really is no reason I can see as to why I wouldn't have just defined fn_gfs_ctl to include INPUT/ prepended.  I will address this in the next commit.

bensonr commented on tools/external_ic.F90:
```diff
@@ -1983,19 +2038,27 @@ subroutine get_ecmwf_ic( Atm, fv_domain )
       allocate (ak_gfs(levp_gfs+1))
       allocate (bk_gfs(levp_gfs+1))
-      call read_data('INPUT/'//trim(fn_gfs_ctl),'vcoord',wk2, no_domain=.TRUE.)
+      fname = 'INPUT/'//trim(fn_gfs_ctl)
```
same comment.

bensonr commented on tools/external_ic.F90:
```diff
@@ -1006,48 +1023,26 @@ subroutine get_hrrr_ic (Atm, fv_domain)
-      endif
-      call mpp_error(NOTE,'==> External_ic::get_hrrr_ic: using control file '//trim(fn_hrr_ctl)//' for HRRR IC')
-
+      fname = 'INPUT/'//trim(fn_hrr_ctl)
```
Same comment as with fn_gfs_ctl.

bensonr commented:
@laurenchilutti - do you feel ready for a formal review?

laurenchilutti commented:
@bensonr Yes, I am ready for a formal review.  I updated the description to include testing completed.  Note that I have not been able to test a nested run warm-starting from restarts - I am working with Joseph on this.

bensonr commented:
>
> @bensonr <https://github.com/bensonr> Yes, I am ready for a formal
> review. I updated the description to include testing completed. Note that I
> have not been able to test a nested run warm-starting from restarts - I am
> working with Joseph on this.
>
We can add that as a note to the CHANGELOG.md for the updates.

uramirez8707 commented on GFDL_tools/fv_ada_nudge.F90:
```diff
@@ -1793,26 +1780,29 @@ subroutine fv_ada_nudge_init(time, axes, npz, zvir, ak, bk, ts, phis, gridstruct
+    if (open_file(ada_driver_restart, restart_file, "read", domain, is_restart=.true.)
+      call register_axis(ada_driver_restart, "xaxis_1", size(Atm_var%u_adj,1)
+      call register_axis(ada_driver_restart, "yaxis_1", size(Atm_var%u_adj,2)
+      call register_axis(ada_driver_restart, "zaxis_1", size(Atm_var%u_adj,3)
```
Is this a domain decomposed file? If it is then you have to replace `size(Atm_var%u_adj,1)` and `size(Atm_var%u_adj,2)` with "x" and "y", otherwise every rank is going to get the same portion of the data.

uramirez8707 commented on GFDL_tools/fv_ada_nudge.F90:
```diff
@@ -1793,26 +1780,29 @@ subroutine fv_ada_nudge_init(time, axes, npz, zvir, ak, bk, ts, phis, gridstruct
+      call register_axis(ada_driver_restart, "zaxis_1", size(Atm_var%u_adj,3)
+      call register_axis(ada_driver_restart, "Time", unlimited)
+      call register_restart_field(ada_driver_restart, &
+           & "u_adj", Atm_var%u_adj(:,:,:), (/"xaxis_1", "yaxis_1", "zaxis_1", "Time"/))
```
I think gnu does not like the different strings lengths in:
```F90
(/"xaxis_1", "yaxis_1", "zaxis_1", "Time"/))
```

laurenchilutti replied:
Yes, I didn't notice this.  Will fix in next commit.

uramirez8707 commented on GFDL_tools/read_climate_nudge_data.F90:
```diff
@@ -152,58 +139,57 @@ subroutine read_climate_nudge_data_init (nlon, nlat, nlev, ntime)
+            endif
+         enddo
+         deallocate(fields)
+         call close_file()
```
Shouldn't this be `close_file(fileobj)` ?

uramirez8707 commented on GFDL_tools/read_climate_nudge_data.F90:
```diff
@@ -152,58 +139,57 @@ subroutine read_climate_nudge_data_init (nlon, nlat, nlev, ntime)
-     enddo
-     deallocate(fields)
-
+     if (open_file(fileobj, trim(filenames(n)), "read")) then
```
Can you also add `pelist=pes` argument to the open_file here?

uramirez8707 commented on GFDL_tools/read_climate_nudge_data.F90:
```diff
@@ -249,16 +234,22 @@ subroutine read_time ( times, units, calendar )
+             time_axis = Files(n)%axes(INDEX_TIME)
+             call get_variable_units(fileobj, time_axis, units)
+             if( variable_att_exists(fileobj, time_axis, calendar) then
+                 call get_time_calendar(file_obj, time_axis, calendar)
```
should this be `fileobj` instead of `file_obj`?

uramirez8707 commented on GFDL_tools/read_climate_nudge_data.F90:
```diff
@@ -281,30 +272,32 @@ subroutine read_grid ( lon, lat, ak, bk )
-         ak = ak*pref
-      else
-         ak = 0.
+      if (open_file(fileobj, trim(filenames(1)), "read")) then
```
add pelist argument to open_file call

uramirez8707 commented on GFDL_tools/read_climate_nudge_data.F90:
```diff
@@ -428,7 +421,10 @@ subroutine read_climate_nudge_data_2d (itime, field, dat, is, js)
      nread(2) = size(dat,2)
 
-     call mpp_read(Files(n)%ncid, Files(n)%fields(this_index), dat, start, nread)
+     if (open_file(fileobj, trim(filenames(n)), "read")) then
```
add pelist argument to open_file call

uramirez8707 commented on GFDL_tools/read_climate_nudge_data.F90:
```diff
@@ -505,7 +501,10 @@ subroutine read_climate_nudge_data_3d (itime, field, dat, is, js)
      nread(3) = size(dat,3)
 
-     call mpp_read(Files(n)%ncid, Files(n)%fields(this_index), dat, start, nread)
+     if (open_file(fileobj, trim(filenames(n)), "read")) then
```
add pelist argument to open_file call

uramirez8707 commented on tools/external_ic.F90:
```diff
@@ -253,13 +253,13 @@ subroutine get_cubed_sphere_terrain( Atm, fv_domain )
     tile_id = mpp_get_tile_id( fv_domain )
 
-       call get_tile_string(fname, 'INPUT/fv_core.res'//trim(gn)//'.tile', tile_id(n), '.nc' )
+    fname = 'INPUT/fv_core.res'//trim(gn)//'.nc'
```
I *think* you can get rid of the '//trim(gn)//', this should already be added by fms2_io.

laurenchilutti replied:
I tested removing this and answers reproduce.

uramirez8707 commented on tools/external_ic.F90:
```diff
@@ -396,22 +394,15 @@ subroutine get_nggps_ic (Atm, fv_domain)
-!--- set the 'nestXX' appendix for all files using fms_io
-    if (Atm%grid_number > 1) then
-       write(gn,'(A4, I2.2)') "nest", Atm%grid_number
+    if( open_file(Gfs_ctl, fn_gfs_ctl, "read") ) then
```
you should add the pelist argument to the call.

uramirez8707 commented on tools/external_ic.F90:
```diff
@@ -2474,8 +2531,8 @@ subroutine get_fv_ic( Atm, fv_domain, nq )
 
-      if( file_exist(fname) ) then
-          call field_size(fname, 'T', tsize, field_found=found)
+      if( file_exists(fname) .and. open_file(Latlon_dyn, fname, "read", is_restart=.true.) ) then
```
I think you can get rid of the `file_exists`, if the file does not exists then open_file is never going to be true.

uramirez8707 commented on tools/external_ic.F90:
```diff
@@ -2529,20 +2586,21 @@ subroutine get_fv_ic( Atm, fv_domain, nq )
       fname = Atm%flagstruct%res_latlon_tracers
 
-      if( file_exist(fname) ) then
+      if( file_exists(fname) .and. open_file(Latlon_tra, fname, "read", is_restart=.true.) ) then
```
Same as above.

uramirez8707 commented on tools/fv_restart.F90:
```diff
@@ -167,25 +169,28 @@ subroutine fv_restart(fv_domain, Atm, dt_atmos, seconds, days, cold_start, grid_
        else
           fname='INPUT/fv_core.res.nc'
-          do_read_restart = file_exist('INPUT/fv_core.res.nc') .or. file_exist('INPUT/fv_core.res.tile1.nc')
+          do_read_restart = file_exists('INPUT/fv_core.res.nc') .or. file_exists('INPUT/fv_core.res.tile1.nc')
```
I am not sure about this. 
If you are doing an ensemble run (i.e SPEAR), the restarts are named "INPUT/fv_core.ensXX.res.nc", but `file_exists` is only going to check for 'INPUT/fv_core.res.nc', so do_read_restart is going to be set to false...

laurenchilutti replied:
@uramirez8707 Would an fms2io open_file() call  adequately check for .ensXX files?

uramirez8707 replied:
yes I think that could work, with is_restart=.true.  and you can use the `FmsNetcdfFile_t` fileobj

uramirez8707 commented on tools/fv_io.F90:
```diff
@@ -590,34 +712,117 @@ subroutine  fv_io_write_restart(Atm, timestamp)
+    else
+      fname = 'RESTART/fv_core.res'//trim(suffix)//'.nc'
+    endif
+    Atm%Fv_restart_is_open = open_file(Atm%Fv_restart, fname, "overwrite", is_restart=.true.)
```
With this every rank is going to write the restart, so add the pelist argument to the open_file call.

laurenchilutti commented on model/fv_regional_bc.F90:
```diff
@@ -6064,32 +6072,47 @@ subroutine dump_field_3d (domain, name, field, isd, ied, jsd, jed, nlev, stag)
-
-    call mpp_write_meta( unit, x, 'grid_xt', 'km', 'X distance', 'X', domain=xdom, data=(/(i*1.0,i=1,nxg)/) )
-    call mpp_write_meta( unit, y, 'grid_yt', 'km', 'Y distance', 'Y', domain=ydom, data=(/(j*1.0,j=1,nyg)/) )
-    call mpp_write_meta( unit, z, 'lev',     'km', 'Z distance',                   data=(/(i*1.0,i=1,nz)/) )
```
I added the write_data call in the latest commit

uramirez8707 reviewed (approved):
Thanks @laurenchilutti. It all looks good to me.

lharris4 reviewed (commented):
This looks good.  I am really glad we can get rid of the INTERNAL_FILE_NML. One question: has nggps_ic been tested with these changes?

bensonr commented:
@lharris4 - there is a "How this has been tested" section in the description of the PR (thanks to the PR template we appropriated from the NOAA-GFDL/FMS project).  Please let us know if there is untested functionality that needs to be addressed.

bensonr commented on GFDL_tools/fv_ada_nudge.F90:
```diff
@@ -1793,26 +1782,35 @@ subroutine fv_ada_nudge_init(time, axes, npz, zvir, ak, bk, ts, phis, gridstruct
+    if (open_file(ada_driver_restart, restart_file, "read", domain, is_restart=.true.)
+      call register_axis(ada_driver_restart, "xaxis_1", "X")
+      call register_axis(ada_driver_restart, "yaxis_1", "Y")
+      call register_axis(ada_driver_restart, "zaxis_1", size(Atm_var%u_adj,3)
```
I think you are missing a final parenthesis here.  Also, why is the third argument an integer here yet in the two calls above it is a character?

uramirez8707 replied:
The `"x"/"y"` indicate that dimension is domain decomposed in x/y. 
The `size(Atm_var%u_adj,3)` indicates the size of that dimension.

bensonr commented on GFDL_tools/fv_ada_nudge.F90:
```diff
@@ -2516,7 +2514,21 @@ subroutine fv_ada_nudge_end
 
-    call save_restart(ada_driver_restart)
+    if (open_file(ada_driver_restart, restart_file, "overwrite", domain, is_restart=.true.) then
+       call fv_io_register_axis(ada_driver_restart, numx=1, numy=1, numz=1, zsize=(/size(Atm_var%u_adj,3)/))
```
Why use fv_io_register axis here, but in other cases you use the direct call to register_axis?

laurenchilutti replied:
I tried to use fv_io_register_axis when I could to limit the amount of duplicate code needed.  I can go through and ensure that I use the fv_io_register_axis wherever possible.

bensonr commented on GFDL_tools/fv_diag_column.F90:
```diff
@@ -394,10 +393,7 @@ subroutine debug_column(pt, delp, delz, u, v, w, q, npz, ncnst, sphum, nwat, zvi
-       call mpp_flush(unit)
-
-
+ 
```
You can still use the fortran flush in place of the mpp_flush.

bensonr reviewed (commented):
I believe you  still have an instance of INTERNAL_FILE_NML in fv_diag_column.F90
